soco/soap.py

```diff
@@ -0,0 +1,263 @@
+# appears to use basic SOAP features, and after experimenting with other
+# libraries, it seems best to write our own.
+
+# Some is the same as that in services.py.
+# TODO: refactor services.py to depend on this code
+
+
+from __future__ import unicode_literals, absolute_import
+
+import logging
+from xml.sax.saxutils import escape
+
+import requests
+
+from soco.utils import prettify
+from soco.xml import XML
+from soco.exceptions import SoCoException
+
+_LOG = logging.getLogger(__name__)
+
+
+class SoapFault(SoCoException):
+
+    """ An exception encapsulating a SOAP Fault. """
+
+    def __init__(self, faultcode, faultstring, detail=None):
+        """ Args:
+                faultcode (str): The SOAP faultcode
+                faultstring (str): The SOAP faultstring
+                detail (Element): The SOAP fault detail, as an ElementTree
+                     Element. Default, None
+
+        """
+        self.faultcode = faultcode
+        self.faultstring = faultstring
+        self.detail = detail
+        self.detail_string = XML.tostring(detail) if detail is not None else ''
+        super(SoapFault, self).__init__(faultcode, faultstring)
+
+    def __str__(self):
+        return '%s: %s' % (self.faultcode, self.faultstring)
+
+    def __repr__(self):
+        return "SoapFault(faultcode={0}, faultstring={1}, detail={2})".format(
+            repr(self.faultcode),
+            repr(self.faultstring),
+            repr(self.detail)
+        )
+
+
+# Sonos uses SOAP to send commands in the RPC form. A complete RPC SOAP
+# message should look something like this. See generally
+# http://www.w3.org/TR/2000/NOTE-SOAP-20000508/
+
+# POST Endpoint URL HTTP/1.1
+# HOST: Host of Endpoint URL:port
+# CONTENT-LENGTH: bytes in body
+# CONTENT-TYPE: text/xml; charset="utf-8"
+# SOAPACTION: URI
+#
+# <?xml version="1.0"?>
+# <s:Envelope
+#   xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"
+#   s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
+#   <s:Header>
+#       </Header elements go here>
+#   </s:Header>
+#   <s:Body>
+#       <ns:MethodName xmlns:ns="MethodNamespace>"
+#           <param1>value</param1>
+#           ...
+#           <param_n>value</param_n>
+#       </ns:MethodName>
+#   </s:Body>
+# </s:Envelope>
+
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+# pylint: disable=no-self-use, fixme
+
+
+class SoapMessage(object):
+
+    """ A SOAP Message representing a remote procedure call.
+
+    Uses the Requests library for communication with a SOAP server."""
+
+    def __init__(self, endpoint, method, parameters=None, http_headers=None,
+                 soap_action=None, soap_header=None, namespace=None,
+                 **request_args):
+        """ Args:
+                endpoint (str): The SOAP endpoint URL for this client
+                method (str): The name of the method to call
+                parameters (list): A list of (name, value) tuples containing
```

Comment:
Yes, we could use an ordereddict here, though you have to pass it a list of tuples in order to create it like this:

```OrderedDict([('param1': 1), ('param2': 2)])```

instead of 

``` OrderedDict({'Param1': 1, 'param2': 2}) ```

since the latter does not (necessarily) retain the right order.

So, I'm not sure it helps at all! (And this is why we ended up using a list of tuples in the services code).